Alloy/Renderers/PointsRenderer.swift

```diff
@@ -0,0 +1,166 @@
+
+        if let colorAttachmentsDescriptor = renderPipelineDescriptor.colorAttachments[0] {
+            colorAttachmentsDescriptor.pixelFormat = pixelFormat
+
+            colorAttachmentsDescriptor.isBlendingEnabled = true
+
+            colorAttachmentsDescriptor.rgbBlendOperation = .add
+            colorAttachmentsDescriptor.sourceRGBBlendFactor = .sourceAlpha
+            colorAttachmentsDescriptor.destinationRGBBlendFactor = .oneMinusSourceAlpha
+
+            colorAttachmentsDescriptor.alphaBlendOperation = .add
+            colorAttachmentsDescriptor.sourceAlphaBlendFactor = .sourceAlpha
+            colorAttachmentsDescriptor.destinationAlphaBlendFactor = .oneMinusSourceAlpha
+        }
+
+        self.renderPipelineState = try? context.device
+            .makeRenderPipelineState(descriptor: renderPipelineDescriptor)
+    }
+
+    private static let vertexFunctionName = "pointVertex"
+    private static let fragmentFunctionName = "pointFragment"
+
+}
+
+@available(iOS 11.3, tvOS 11.3, macOS 10.13, *)
+extension PointsRenderer {
+
+    /// Draw a rectangle in a target texture.
+    ///
+    /// - Parameters:
+    ///   - renderPassDescriptor: render pass descriptor to be used.
+    ///   - commandBuffer: command buffer to put the rendering work items into.
+    public func draw(renderPassDescriptor: MTLRenderPassDescriptor,
+                     commandBuffer: MTLCommandBuffer) throws {
+        #if DEBUG
+        // Check render target.
+        guard let renderTarget = renderPassDescriptor.colorAttachments[0].texture
+        else { throw Errors.missingRenderTarget }
+        guard renderTarget.usage.contains(.renderTarget)
+        else { throw Errors.wrongRenderTargetTextureUsage }
+        #endif
+
+        self.renderTargetAspectRatio =
+            Float(renderTarget.width) / Float(renderTarget.height)
+
+        // Draw.
+        commandBuffer.render(descriptor: renderPassDescriptor) { renderEncoder in
+            self.draw(using: renderEncoder)
+        }
+    }
+
+    /// Draw a rectangle in a target texture.
+    ///
+    /// - Parameters:
+    ///   - renderEncoder: container to put the rendering work into.
+    public func draw(using renderEncoder: MTLRenderCommandEncoder) {
+
+        // Push a debug group allowing us to identify render commands in the GPU Frame Capture tool.
+        renderEncoder.pushDebugGroup("Draw Points Geometry")
+
+        // Set render command encoder state.
+        renderEncoder.setRenderPipelineState(self.renderPipelineState)
+        // Set any buffers fed into our render pipeline.
+        renderEncoder.setVertexBuffer(self.pointsBuffer,
+                                      offset: 0,
+                                      index: 0)
+
+        renderEncoder.setFragmentBytes(&self.color,
+                                       length: MemoryLayout<vector_float4>.stride,
+                                       index: 0)
+
+        // Draw.
+        if let pointsBuffer = self.pointsBuffer {
```

Comment:
we should check it earlier + we really don't want to do anything if point count is 0